orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -798,6 +800,167 @@ class PipelineSplitter {
   }
 };
 
+// TODO: move to a proper location for pipeline training
+
+// pipeline plan for each batch
+struct PipelineBatchInfo {
+  // Event pairs for each pipeline slot to WaitEvent when start, and RecordEvent when end
+  std::vector<std::pair<int64_t, int64_t>> events;
+  // indices of retired batches, so their data could be reused
+  // a batch can only be retired after finished backward in stage 0
+  // this can be used to join worker threads or reuse buffers
+  // for example, in a node with N GPUs and B batches to run in pipeline
+  // there will be (N * B) threads created, and by being able to retire,
+  // only (N * (2*N-1)) concurrent threads are needed.
+  // for big number of B, this would be helpful
+  std::vector<int64_t> retired_batches;
+};
+
+class PipelineTimeline {
+ public:
+  struct Slot {
+    enum class Type {
+      Forward,
+      Backward
+    };
+    Type type;
+    size_t batch_id;
+    bool occupied;
+
+    Slot() : occupied(false) {}
+  };
+
+  PipelineTimeline() = default;
+
+  void Initialize(size_t num_stages, size_t num_slots) {
+    slots_.resize(num_stages);
+    for (size_t s = 0; s < num_stages; ++s) {
+      slots_[s].resize(num_slots);
+    }
+  }
+
+  bool IsOccupied(size_t s, size_t t) const {
+    return slots_[s][t].occupied;
+  }
+
+  const Slot& Get(size_t s, size_t t) const {
+    return slots_[s][t];
+  }
+
+  size_t GetNumSlots() const {
+    return slots_[0].size();
+  }
+
+  void Occupy(size_t s, size_t t, size_t batch_id, Slot::Type st) {
+    Slot& slot = slots_[s][t];
+    slot.occupied = true;
+    slot.type = st;
+    slot.batch_id = batch_id;
+  }
+
+ private:
+  std::vector<std::vector<Slot>> slots_;
+};
+
+// pipeline planner for batches
+class PipelineBatchPlanner {
+ private:
+  int64_t max_id_;
+  PipelineTimeline timeline_;
+
+ public:
+  PipelineBatchPlanner()
+      : max_id_(::onnxruntime::contrib::OrtEventPool::GetPoolSize() - 1) {
+  }
+
+  // Generate timeline for one-forward-one-backward scheduling,
+  // which schedules execution in batch order to minimize latency for onging batches
+  // each stage requires 2 pair of wait/record events for FW/BW
+  // except for last stage, which only need one pair of wait/record
+  void GenerateOneFWOneBWTimeline(size_t num_stages, size_t num_batches) {
+    // The first batch has 2 * (num_stages - 1) gaps between FW and BW
+    // then 2 slots for FW/BW in each batch
+    size_t num_slots = 2 * (num_stages - 1) + num_batches * 2;
+    timeline_.Initialize(num_stages, num_slots);
+
+    // fw time slot to start the search for empty ones in each stage
+    std::vector<size_t> t_fw(num_stages);
+    // bw time slot to start the search for empty ones in each stage
+    std::vector<size_t> t_bw(num_stages);
+    for (size_t s = 0; s < num_stages; ++s) {
+      t_fw[s] = 0;
+      t_bw[s] = 0;
+    }
```

Comment:
Can be simplified by:
  std::vector<size_t> t_fw(num_stages, 0);
  std::vector<size_t> t_bw(num_stages, 0);